File: apps/ios/Packages/StoryArcEpub/Sources/EpubReaderFeature/ReadAloudBar.swift
internal import SwiftUI

internal import DesignSystem

/// The transport, on screen only while a book is being read aloud.
///
/// Four controls and no scrubber. `ebook-reader` asks the *lock screen* for "play, pause,
/// and sentence skip", and the reader looking at the page gets the same four plus a stop —
/// because the way out of speech is not obvious from a pause button, and a reader who is
/// done listening should not have to guess.
///
/// It appears where the return control does, above the percentage, and for the same
/// reason: it comes and goes, and a control that appeared among the ones at the top would
/// move them.
///
/// Four glass buttons and no glass behind them. The first version put the row on a
/// `storyArcGlass()` surface and the four on top of it, and on a simulator three of the
/// four glyphs vanished: glass over glass lightens twice, and a tinted symbol on the
/// result is not a symbol any more. These sit directly in the chrome's own
/// `GlassEffectContainer`, exactly as the four at the top of the screen do.
///
/// The same session is drawn outside the reader by ``ReadAloudDock``, and that one is
/// plain rather than glass for the same reason in reverse: the slot it sits in is already
/// the material. Two views, because that difference is real; one ``ReadAloudControl``, so
/// the verbs behind them cannot drift.
///
/// Android's `ReadAloudBar` is the same five decisions in Compose.
struct ReadAloudBar: View {
    @Environment(\.theme) private var theme

    let isSpeaking: Bool
    let onPrevious: () -> Void
    let onToggle: () -> Void
    let onNext: () -> Void
    let onStop: () -> Void

    var body: some View {
        HStack(spacing: StoryArcSpace.sm) {
            control(.previous, action: onPrevious)
            control(.toggle(isSpeaking: isSpeaking), tint: theme.accent, action: onToggle)
            control(.next, action: onNext)
            control(.stop, action: onStop)
        }
    }

    private func control(
        _ control: ReadAloudControl,
        tint: Color? = nil,
        action: @escaping () -> Void
    ) -> some View {
        Button(action: action) {
            Label {
                Text(control.label, bundle: .module)
            } icon: {
                Image(systemName: control.symbol)
            }
            .labelStyle(.iconOnly)
        }
        .buttonStyle(.glass)
        .tint(tint ?? theme.palette.textPrimary)
    }
}

File: apps/android/feature/epubreader/src/main/kotlin/app/storyarc/feature/epubreader/ReadAloudBar.kt
package app.storyarc.feature.epubreader

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Pause
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.SkipNext
import androidx.compose.material.icons.filled.SkipPrevious
import androidx.compose.material.icons.filled.Stop
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Surface
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.res.stringResource
import app.storyarc.core.designsystem.theme.LocalStoryArcPalette
import app.storyarc.core.designsystem.tokens.StoryArcRadius
import app.storyarc.core.designsystem.tokens.StoryArcSpace

/**
 * The transport, on screen only while a book is being read aloud.
 *
 * Four controls and no scrubber. `ebook-reader` asks the *lock screen* for "play, pause,
 * and sentence skip", and the reader looking at the page gets the same three plus a stop —
 * because the way out of speech is not obvious from a pause button, and a reader who is
 * done listening should not have to guess.
 *
 * It sits where the return control does, above the percentage, and for the same reason: it
 * comes and goes, and a control that appeared among the ones at the top would move them.
 *
 * iOS's `ReadAloudBar` is the same five decisions in SwiftUI.
 */
@Composable
internal fun ReadAloudBar(
    isSpeaking: Boolean,
    onPrevious: () -> Unit,
    onToggle: () -> Unit,
    onNext: () -> Unit,
    onStop: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val palette = LocalStoryArcPalette.current

    Surface(
        modifier = modifier,
        shape = RoundedCornerShape(StoryArcRadius.lg),
        color = palette.surfaceRaised,
    ) {
        Row(
            modifier = Modifier.padding(horizontal = StoryArcSpace.xs),
            horizontalArrangement = Arrangement.spacedBy(StoryArcSpace.xs),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Control(
                Icons.Filled.SkipPrevious,
                R.string.readaloud_previous,
                palette.textPrimary,
                onPrevious,
            )
            Control(
                if (isSpeaking) Icons.Filled.Pause else Icons.Filled.PlayArrow,
                if (isSpeaking) R.string.readaloud_pause else R.string.readaloud_play,
                palette.accent,
                onToggle,
            )
            Control(Icons.Filled.SkipNext, R.string.readaloud_next, palette.textPrimary, onNext)
            Control(Icons.Filled.Stop, R.string.readaloud_stop, palette.textPrimary, onStop)
        }
    }
}

@Composable
private fun Control(
    icon: ImageVector,
    label: Int,
    tint: Color,
    onClick: () -> Unit,
) {
    IconButton(onClick = onClick) {
        Icon(imageVector = icon, contentDescription = stringResource(label), tint = tint)
    }
}
